SearchEngine auto-completion suggestions: Prevent extra HTTP requests

b0o opened the pull request:
Previously, when entering text into an OmniBar SearchEngine with a suggestion URL, an HTTP request would be dispatched for each keystroke. When typing long queries, this could lead to rate-limiting, extra network traffic, and is unfriendly to API servers.

---

I've updated the SearchEngine suggestion logic to use a timeout of 200ms
before dispatching an HTTP request. If another keystroke occurs before
this 200ms timeout, the previous request is canceled.

Example Result - When typing "testing" into a SearchEngine dialog with less than 200ms delay between keystrokes:

Old: 7 HTTP requests are dispatched
New: 1 HTTP request is dispatched

Note: It may be desirable to make the timeout duration configurable.
Currently it's hard-coded as 200ms.

---

This has proven to be very beneficial for me.

I've added a lot of custom searchAliasX's to my configuration, and many APIs (including GitHub's) will rate limit you very quickly using the old code.

Also, if an API is very slow to respond, it can cause odd behavior in the OmniBar when several requests finish at once or out of order.

brookhong commented:
Nice, I like it.

Could you please also add some variable(such as `omnibarSuggestionTimeout`) under [runtime](https://github.com/brookhong/Surfingkeys/blob/master/content_scripts/runtime.js#L9) for configuration? So that users could customzie their own timeout with:

    settings.omnibarSuggestionTimeout = 500;

And it would be better if you could update the [readme doc](https://github.com/brookhong/Surfingkeys#properties-list) for this setting as well.

brookhong commented on pages/omnibar.js:
```diff
@@ -950,18 +952,26 @@ var SearchEngine = (function() {
+            return;
+        }
+        var val = Omnibar.input.val();
+        if(_pendingRequest) {
```
Miss a space between `if` and `(`.

brookhong commented on pages/omnibar.js:
```diff
@@ -950,18 +952,26 @@ var SearchEngine = (function() {
+                self.listSuggestion(resp);
             });
-        }
+        }, 200);
```
You could use `runtime.conf.omnibarSuggestionTimeout` instead of 200 here.

b0o commented:
@brookhong  Thanks for the feedback!

I've updated my PR to use `runtime.conf.omnibarSuggestionTimeout` as suggested. 

I've also updated README.md to reflect this. However, I have not updated README_CN.md as I don't speak Chinese. Perhaps you can help me with the translation?

brookhong commented:
Thanks, merged.

I'll update README_CN.md.
